documents/coredocument.go

```diff
@@ -428,53 +423,68 @@ func (cd *CoreDocument) documentTree(docType string) (tree *proofs.DocumentTree,
 
 }
 
-// GetCollaborators returns the collaborators from the role with READ_SIGN ability.
-func (cd *CoreDocument) GetCollaborators(filterIDs ...identity.DID) (ids []identity.DID, err error) {
-	exclude := make(map[string]struct{})
-	for _, id := range filterIDs {
-		exclude[id.String()] = struct{}{}
+// GetSignCollaborators returns the collaborators excluding the filteredIDs
+// returns collaborators with Read_Sign permissions.
+func (cd *CoreDocument) GetSignCollaborators(filterIDs ...identity.DID) ([]identity.DID, error) {
+	cs, err := cd.getCollaborators(coredocumentpb.Action_ACTION_READ_SIGN)
+	if err != nil {
+		return nil, err
 	}
 
-	for _, c := range cd.Document.Collaborators {
-		id := identity.NewDIDFromBytes(c)
-		if _, ok := exclude[id.String()]; ok {
-			continue
-		}
+	return filterCollaborators(cs, filterIDs...), nil
+}
 
-		ids = append(ids, id)
+// GetCollaborators returns the collaborators excluding the filteredIDs
+// returns collaborators with Read and Read_Sign permissions.
+func (cd *CoreDocument) GetCollaborators(filterIDs ...identity.DID) ([]identity.DID, error) {
+	cs, err := cd.getCollaborators(coredocumentpb.Action_ACTION_READ_SIGN, coredocumentpb.Action_ACTION_READ)
+	if err != nil {
+		return nil, err
 	}
 
-	return ids, nil
+	return filterCollaborators(cs, filterIDs...), nil
 }
 
-// fetchNewUniqueCollaborators returns the unique collaborators from newCollabs in reference to oldCollabs.
-func fetchNewUniqueCollaborators(oldCollabs [][]byte, newCollabs []string) (ids []identity.DID, err error) {
-	ocsm := make(map[string]struct{})
-	for _, c := range oldCollabs {
-		cs := strings.ToLower(hexutil.Encode(c))
-		ocsm[cs] = struct{}{}
-	}
+// getCollaborators returns all the collaborators who belongs to the actions passed.
+func (cd *CoreDocument) getCollaborators(actions ...coredocumentpb.Action) (ids []identity.DID, err error) {
+	findRole(cd.Document, func(_, _ int, role *coredocumentpb.Role) bool {
+		if len(role.Collaborators) < 1 {
+			return false
+		}
 
-	var uc []string
-	for _, c := range newCollabs {
-		cs := strings.ToLower(c)
-		if _, ok := ocsm[cs]; ok {
-			continue
+		for _, c := range role.Collaborators {
+			// TODO(ved): we should ideally check the address length of 20
+			// we will still keep the error return to the function so that once check is in, we don't have to refactor this function
+			ids = append(ids, identity.NewDIDFromBytes(c))
```

Comment:
Note the check should happen in the `NewDIDFromBytes` functions here, I think its better to add it now, it shouldn't be a major change, isn't it?

Reply:
I would add it as a separate PR, there will be some refactoring involved.